File: apps/cli-py/src/betteruseofai/log.py
"""The event log: the history that outlives the transcripts.

Mirrors apps/cli-ts/src/log.ts line for line in behaviour, and byte for byte in
what it writes. One JSON object per line, keys sorted, no spaces, one file per
month, append only; the last line for an id wins when the log is read, and
prune compacts. Tokens are stored rather than figures so a dataset update
re-prices the whole history.

The location is shared with the TypeScript tool and the plugin's Stop hook on
purpose: under the Claude configuration directory, whichever tool is writing,
with BUAI_LOG_DIR as the override.
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .estimate import TokenCounts, UsageEvent

LOG_VERSION = 1
LOG_WARN_BYTES = 200 * 1024 * 1024

_MONTH_FILE = re.compile(r"^\d{4}-\d{2}\.jsonl$")
_MONTH = re.compile(r"^\d{4}-\d{2}$")


def default_log_dir() -> str:
    override = os.environ.get("BUAI_LOG_DIR")
    if override:
        return override
    base = os.environ.get("CLAUDE_CONFIG_DIR") or str(Path.home() / ".claude")
    return str(Path(base) / "betteruseofai" / "log")


def default_out_dir(log_dir: str) -> str:
    """The dashboard file goes beside the log rather than in it."""
    return str(Path(log_dir).parent)


def to_line(event: UsageEvent, recorded: str) -> dict[str, Any]:
    tokens = event.tokens
    body: dict[str, Any] = {
        "input": tokens.input or 0,
        "output": tokens.output or 0,
        "cachedRead": tokens.cached_read or 0,
        "cachedWrite": tokens.cached_write or 0,
    }
    # A number is a count, None is undisclosed, and "absent" means the model
    # cannot think, which is written as no key at all.
    if tokens.thinking != "absent":
        body["thinking"] = tokens.thinking
    if tokens.tool:
        body["tool"] = tokens.tool
    if tokens.estimated:
        body["estimated"] = True
        body["estimator"] = tokens.estimator

    line: dict[str, Any] = {
        "v": LOG_VERSION,
        "id": event.id,
        "ts": event.timestamp,
        "recorded": recorded,
        "surface": event.surface,
        "hosting": event.hosting,
        "model": event.model_raw,
        "tokens": body,
    }
    if event.session_id:
        line["session"] = event.session_id
    if event.region_hint:
        line["region"] = event.region_hint
    project = event.meta.get("project")
    if isinstance(project, str) and project:
        line["project"] = project
    branch = event.meta.get("branch")
    if isinstance(branch, str) and branch:
        line["branch"] = branch
    return line


def serialise(line: dict[str, Any]) -> str:
    return json.dumps(line, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def from_line(line: dict[str, Any], file: str) -> UsageEvent:
    body = line["tokens"]
    tokens = TokenCounts(
        input=body.get("input", 0),
        output=body.get("output", 0),
        cached_read=body.get("cachedRead", 0),
        cached_write=body.get("cachedWrite", 0),
        thinking=body.get("thinking", "absent"),
        tool=body.get("tool", 0),
        estimated=body.get("estimated") is True,
        estimator=body.get("estimator", "provider"),
    )
    meta: dict[str, Any] = {"file": file, "fromLog": True}
    if line.get("project"):
        meta["project"] = line["project"]
    if line.get("branch"):
        meta["branch"] = line["branch"]
    return UsageEvent(
        id=line["id"],
        surface=line["surface"],
        hosting=line["hosting"],
        model_raw=line["model"],
        tokens=tokens,
        timestamp=line["ts"],
        session_id=line.get("session"),
        region_hint=line.get("region"),
        meta=meta,
    )


def _is_line(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and value.get("v") == LOG_VERSION
        and isinstance(value.get("id"), str)
        and isinstance(value.get("ts"), str)
        and isinstance(value.get("tokens"), dict)
    )


def _month_files(directory: str) -> list[Path]:
    root = Path(directory)
    if not root.exists():
        return []
    return sorted(p for p in root.iterdir() if p.is_file() and _MONTH_FILE.match(p.name))


@dataclass
class LogRead:
    events: list[UsageEvent]
    duplicates: int
    skipped: int
    files: int
    bytes: int


def read_log(directory: str) -> LogRead:
    by_id: dict[str, UsageEvent] = {}
    duplicates = 0
    skipped = 0
    size = 0
    files = _month_files(directory)
    for file in files:
        text = file.read_text(encoding="utf-8")
        size += len(text.encode("utf-8"))
        for raw in text.split("\n"):
            trimmed = raw.strip()
            if not trimmed:
                continue
            try:
                value = json.loads(trimmed)
            except json.JSONDecodeError:
                skipped += 1
                continue
            if not _is_line(value):
                skipped += 1
                continue
            if value["id"] in by_id:
                duplicates += 1
            by_id[value["id"]] = from_line(value, str(file))
    events = sorted(by_id.values(), key=lambda event: (event.timestamp, event.id))
    return LogRead(events, duplicates, skipped, len(files), size)


def append_log(directory: str, events: list[UsageEvent], recorded: str) -> int:
    if not events:
        return 0
    root = Path(directory)
    root.mkdir(parents=True, exist_ok=True)
    by_month: dict[str, list[str]] = {}
    for event in events:
        month = event.timestamp[:7]
        if not _MONTH.match(month):
            continue
        by_month.setdefault(month, []).append(serialise(to_line(event, recorded)))
    written = 0
    for month in sorted(by_month):
        lines = by_month[month]
        with (root / f"{month}.jsonl").open("a", encoding="utf-8", newline="\n") as handle:
            handle.write("\n".join(lines) + "\n")
        written += len(lines)
    return written


def log_size(directory: str) -> tuple[int, int]:
    files = _month_files(directory)
    return len(files), sum(f.stat().st_size for f in files)


@dataclass
class PruneResult:
    removed: int
    duplicates: int
    kept: int
    files_before: int
    files_after: int
    bytes_before: int
    bytes_after: int


def prune_log(directory: str, before: str | None, dry_run: bool) -> PruneResult:
    files = _month_files(directory)
    cutoff = before or ""
    removed = 0
    duplicates = 0
    kept = 0
    bytes_before = 0
    bytes_after = 0
    files_after = 0

    for file in files:
        text = file.read_text(encoding="utf-8")
        bytes_before += len(text.encode("utf-8"))
        last: dict[str, dict[str, Any]] = {}
        for raw in text.split("\n"):
            trimmed = raw.strip()
            if not trimmed:
                continue
            try:
                value = json.loads(trimmed)
            except json.JSONDecodeError:
                continue
            if not _is_line(value):
                continue
            if value["id"] in last:
                duplicates += 1
            last[value["id"]] = value
        survivors = []
        for line in last.values():
            if cutoff == "" or line["ts"] >= cutoff:
                survivors.append(line)
            else:
                removed += 1
        kept += len(survivors)
        if survivors:
            out = "\n".join(serialise(line) for line in survivors) + "\n"
            bytes_after += len(out.encode("utf-8"))
            files_after += 1
            if not dry_run:
                file.write_text(out, encoding="utf-8", newline="\n")
        elif not dry_run:
            file.unlink()

    return PruneResult(
        removed, duplicates, kept, len(files), files_after, bytes_before, bytes_after
    )

File: apps/cli-ts/src/log.ts
import { appendFileSync, existsSync, mkdirSync, readdirSync, readFileSync, rmSync, statSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

import type { TokenCounts, UsageEvent } from '@betteruseofai/core';

/**
 * The event log: the history that outlives the transcripts.
 *
 * Claude Code deletes its transcripts after a retention period, thirty days by
 * default, and Codex keeps its own. Everything the tools show is re-priced
 * from those files, so the moment a transcript goes the turns in it vanish
 * from every total. This log is the fix. Both the Stop hook in the plugin and
 * every run of the command line tool append the turns they read here, and the
 * dashboard reads it back beside whatever transcripts still exist.
 *
 * Three rules shape the format.
 *
 * Tokens are stored, never figures. A dataset update then re-prices the whole
 * history rather than freezing the numbers a turn was priced at on the day.
 *
 * One JSON object per line, keys sorted, no spaces, one file per month. It is
 * the same line whether the TypeScript tool or the Python one wrote it, so the
 * two can share a log, and the parity harness compares the bytes.
 *
 * Append only. A turn that appears twice, because a message was revised in a
 * later chunk or two writers overlapped, is resolved when the log is read:
 * the last line for an id wins. Nothing has to read the log to write to it,
 * which is what lets a hook with a ten second budget append and leave. The
 * prune command compacts.
 *
 * Where it lives is deliberate. The plugin's per-session caches sit under the
 * plugin's own data directory, which Claude Code hands to hooks and nobody
 * else. A log that two different writers must find has to be somewhere both
 * agree on, so it is under the Claude configuration directory regardless of
 * which tool is writing, and BUAI_LOG_DIR overrides it for tests and for
 * anyone who wants it elsewhere.
 */

export const LOG_VERSION = 1;

export interface LogLine {
  v: number;
  id: string;
  ts: string;
  recorded: string;
  surface: string;
  hosting: 'cloud' | 'local';
  model: string;
  session?: string;
  region?: string;
  project?: string;
  branch?: string;
  tokens: {
    input: number;
    output: number;
    cachedRead: number;
    cachedWrite: number;
    /** A number is a count, null is undisclosed, and absent means the model cannot think. */
    thinking?: number | null;
    tool?: number;
    estimated?: true;
    estimator?: string;
  };
}

export const defaultLogDir = (): string =>
  process.env['BUAI_LOG_DIR'] ??
  join(process.env['CLAUDE_CONFIG_DIR'] ?? join(homedir(), '.claude'), 'betteruseofai', 'log');

/** The directory the dashboard file is written to, beside the log rather than in it. */
export const defaultOutDir = (logDir: string): string => join(logDir, '..');

const monthOf = (timestamp: string): string => timestamp.slice(0, 7);

/** Keys sorted by code point, no spaces: the same bytes the Python tool writes. */
const sortKeys = (value: unknown): unknown => {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value && typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .filter(([, item]) => item !== undefined)
        .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
        .map(([key, item]) => [key, sortKeys(item)]),
    );
  }
  return value;
};

export const toLine = (event: UsageEvent, recorded: string): LogLine => {
  const tokens = event.tokens;
  const line: LogLine = {
    v: LOG_VERSION,
    id: event.id,
    ts: event.timestamp,
    recorded,
    surface: event.surface,
    hosting: event.hosting,
    model: event.modelRaw,
    tokens: {
      input: tokens.input ?? 0,
      output: tokens.output ?? 0,
      cachedRead: tokens.cachedRead ?? 0,
      cachedWrite: tokens.cachedWrite ?? 0,
    },
  };
  if (tokens.thinking !== undefined) line.tokens.thinking = tokens.thinking;
  if (tokens.tool) line.tokens.tool = tokens.tool;
  if (tokens.estimated) {
    line.tokens.estimated = true;
    line.tokens.estimator = tokens.estimator;
  }
  if (event.sessionId) line.session = event.sessionId;
  if (event.regionHint) line.region = event.regionHint;
  const project = event.meta?.['project'];
  if (typeof project === 'string' && project) line.project = project;
  const branch = event.meta?.['branch'];
  if (typeof branch === 'string' && branch) line.branch = branch;
  return line;
};

export const serialise = (line: LogLine): string => JSON.stringify(sortKeys(line));

export const fromLine = (line: LogLine, file: string): UsageEvent => {
  const tokens: TokenCounts = {
    input: line.tokens.input,
    output: line.tokens.output,
    cachedRead: line.tokens.cachedRead,
    cachedWrite: line.tokens.cachedWrite,
    estimated: line.tokens.estimated === true,
    estimator: line.tokens.estimator ?? 'provider',
  };
  if ('thinking' in line.tokens) tokens.thinking = line.tokens.thinking;
  if (line.tokens.tool) tokens.tool = line.tokens.tool;
  return {
    id: line.id,
    surface: line.surface,
    hosting: line.hosting,
    modelRaw: line.model,
    modelId: null,
    tokens,
    timestamp: line.ts,
    ...(line.session ? { sessionId: line.session } : {}),
    ...(line.region ? { regionHint: line.region } : {}),
    meta: {
      file,
      fromLog: true,
      ...(line.project ? { project: line.project } : {}),
      ...(line.branch ? { branch: line.branch } : {}),
    },
  };
};

const isLine = (value: unknown): value is LogLine =>
  typeof value === 'object' &&
  value !== null &&
  (value as LogLine).v === LOG_VERSION &&
  typeof (value as LogLine).id === 'string' &&
  typeof (value as LogLine).ts === 'string' &&
  typeof (value as LogLine).tokens === 'object';

const monthFiles = (dir: string): string[] => {
  if (!existsSync(dir)) return [];
  return readdirSync(dir)
    .filter((name) => /^\d{4}-\d{2}\.jsonl$/.test(name))
    .sort()
    .map((name) => join(dir, name));
};

export interface LogRead {
  events: UsageEvent[];
  /** Lines that were superseded by a later line for the same id. */
  duplicates: number;
  /** Lines that could not be read as a log line and were skipped. */
  skipped: number;
  files: number;
  bytes: number;
}

/** Every event in the log, the last line for an id winning, in timestamp order. */
export const readLog = (dir: string): LogRead => {
  const byId = new Map<string, UsageEvent>();
  let duplicates = 0;
  let skipped = 0;
  let bytes = 0;
  const files = monthFiles(dir);
  for (const file of files) {
    const text = readFileSync(file, 'utf8');
    bytes += Buffer.byteLength(text, 'utf8');
    for (const raw of text.split('\n')) {
      const trimmed = raw.trim();
      if (trimmed === '') continue;
      let value: unknown;
      try {
        value = JSON.parse(trimmed);
      } catch {
        skipped += 1;
        continue;
      }
      if (!isLine(value)) {
        skipped += 1;
        continue;
      }
      if (byId.has(value.id)) duplicates += 1;
      byId.set(value.id, fromLine(value, file));
    }
  }
  const events = [...byId.values()].sort((a, b) =>
    a.timestamp < b.timestamp ? -1 : a.timestamp > b.timestamp ? 1 : a.id < b.id ? -1 : 1,
  );
  return { events, duplicates, skipped, files: files.length, bytes };
};

/** Appends these events, one line each, to the month file their timestamp falls in. */
export const appendLog = (dir: string, events: UsageEvent[], recorded: string): number => {
  if (events.length === 0) return 0;
  mkdirSync(dir, { recursive: true });
  const byMonth = new Map<string, string[]>();
  for (const event of events) {
    const month = monthOf(event.timestamp);
    if (!/^\d{4}-\d{2}$/.test(month)) continue;
    const lines = byMonth.get(month) ?? [];
    lines.push(serialise(toLine(event, recorded)));
    byMonth.set(month, lines);
  }
  let written = 0;
  for (const [month, lines] of [...byMonth.entries()].sort()) {
    appendFileSync(join(dir, `${month}.jsonl`), `${lines.join('\n')}\n`, 'utf8');
    written += lines.length;
  }
  return written;
};

export interface LogSize {
  files: number;
  bytes: number;
}

export const logSize = (dir: string): LogSize => {
  const files = monthFiles(dir);
  return { files: files.length, bytes: files.reduce((total, file) => total + statSync(file).size, 0) };
};

/** Above this the dashboard and the doctor suggest pruning. */
export const LOG_WARN_BYTES = 200 * 1024 * 1024;

export interface PruneResult {
  removed: number;
  duplicates: number;
  kept: number;
  filesBefore: number;
  filesAfter: number;
  bytesBefore: number;
  bytesAfter: number;
}

/**
 * Drops every turn before a moment and compacts what is left, so each id
 * appears once. With dryRun nothing is written and the counts say what would
 * have happened.
 */
export const pruneLog = (dir: string, before: string | null, dryRun: boolean): PruneResult => {
  const files = monthFiles(dir);
  const before_ = before ?? '';
  let removed = 0;
  let duplicates = 0;
  let kept = 0;
  let bytesBefore = 0;
  let bytesAfter = 0;
  let filesAfter = 0;

  for (const file of files) {
    const text = readFileSync(file, 'utf8');
    bytesBefore += Buffer.byteLength(text, 'utf8');
    const last = new Map<string, LogLine>();
    for (const raw of text.split('\n')) {
      const trimmed = raw.trim();
      if (trimmed === '') continue;
      let value: unknown;
      try {
        value = JSON.parse(trimmed);
      } catch {
        continue;
      }
      if (!isLine(value)) continue;
      if (last.has(value.id)) duplicates += 1;
      last.set(value.id, value);
    }
    const survivors = [...last.values()].filter((line) => {
      const keep = before_ === '' || line.ts >= before_;
      if (!keep) removed += 1;
      return keep;
    });
    kept += survivors.length;
    if (survivors.length > 0) {
      const out = `${survivors.map(serialise).join('\n')}\n`;
      bytesAfter += Buffer.byteLength(out, 'utf8');
      filesAfter += 1;
      if (!dryRun) writeFileSync(file, out, 'utf8');
    } else if (!dryRun) {
      rmSync(file);
    }
  }

  return { removed, duplicates, kept, filesBefore: files.length, filesAfter, bytesBefore, bytesAfter };
};
